keel-titus-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/image/CurrentlyDeployedDockerImageApprover.kt

```diff
@@ -0,0 +1,65 @@
+package com.netflix.spinnaker.keel.api.titus.image
+
+import com.netflix.spinnaker.keel.api.ArtifactType.DOCKER
+import com.netflix.spinnaker.keel.api.matchingArtifact
+import com.netflix.spinnaker.keel.api.titus.cluster.TitusClusterSpec
+import com.netflix.spinnaker.keel.docker.ReferenceProvider
+import com.netflix.spinnaker.keel.events.ArtifactVersionDeployed
+import com.netflix.spinnaker.keel.persistence.ArtifactRepository
+import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.persistence.ResourceRepository
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+
+@Component
+class CurrentlyDeployedDockerImageApprover(
+  private val artifactRepository: ArtifactRepository,
+  private val resourceRepository: ResourceRepository,
+  private val deliveryConfigRepository: DeliveryConfigRepository
+) {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  @EventListener(ArtifactVersionDeployed::class)
+  fun onArtifactVersionDeployed(event: ArtifactVersionDeployed) =
+    runBlocking {
+      val resourceId = event.resourceId
+      val resource = resourceRepository.get(resourceId)
+      val deliveryConfig = deliveryConfigRepository.deliveryConfigFor(resourceId)
+      val env = deliveryConfigRepository.environmentFor(resourceId)
+
+      if (resource.spec is TitusClusterSpec) {
+        val spec = resource.spec as TitusClusterSpec // needed because kotlin can't cast it automatically
```

Comment:
If you swapped the statements it would be able to, i.e.

```
val spec = resource.spec
if (spec is TitusClusterSpec) {
```

It can't do it this way because `spec` is defined on an interface so the compiler can't tell what kind of trickery might be involved in returning that value and whether it's guaranteed to be the same every time.

Another way is to use the `as?` operator that will return `null` if the type cast fails:

```
(resource.spec as? TitusClusterSpec)?.let { spec ->
```

Reply:
Nice! I'll do that in the ec2 version as well.